core/src/main/scala/neotypes/Stream.scala

```diff
@@ -1,14 +1,40 @@
 package neotypes
 
+import org.reactivestreams.Publisher
+
+import scala.collection.compat.Factory
+
 @annotation.implicitNotFound("The stream type ${S} is not supported by neotypes")
 trait Stream[S[_]] {
   type F[T]
 
-  def init[T](value: () => F[Option[T]]): S[T]
+  // Legacy module ------------------------------------------------------------
+  private[neotypes] def init[A](value: () => F[Option[A]]): S[A]
+
+  private[neotypes] def onComplete[A](s: S[A])(f: => F[Unit]): S[A]
+
+  private[neotypes] def fToS[A](f: F[S[A]]): S[A]
+  // --------------------------------------------------------------------------
+
+  // New (Rx) module ----------------------------------------------------------
+  private[neotypes] def fromRx[A](publisher: Publisher[A]): S[A]
```

Comment:
Are you leaking the reactor-core Publisher class into the API?  Would it be better to do something similar to the CompletionStageOps which keeps the java CompletionStage out of the API?

Reply:
My thought is to do something like this (exception case is missing)
```java
  implicit class PublisherStageOps[A](private val underlying: Publisher[A]) extends AnyVal {
    private final def acceptImpl[B](cb: Either[Throwable, B] => Unit)
                                   (f: A => Either[Throwable, B])
                                   (g: Throwable => Either[Throwable, B]): Unit = {
      internal.utils.void(
        Flux.from(underlying)
          .doOnError(th => cb(g(th)))
          .doOnNext(a => cb(f(a)))
      )
    }

    def accept[B](cb: Either[Throwable, B] => Unit)
                 (f: A => Either[Throwable, B]): Unit =
      acceptImpl(cb)(f)(defaultExHandler)

    def acceptExceptionally[B](cb: Either[Throwable, B] => Unit)
                              (f: A => Either[Throwable, B])
                              (g: PartialFunction[Throwable, Either[Throwable, B]]): Unit =
      acceptImpl(cb)(f)(g.orElse(defaultExHandler))

    def acceptVoid(cb: Either[Throwable, Unit] => Unit)
                  (implicit ev: A =:= Void): Unit = {
      internal.utils.void(ev)
      acceptImpl(cb)(_ => Right(()))(defaultExHandler)
    }
  }
```

Reply:
@geoffjohn11 technically speaking neither **Async** nor **Stream** are public API, they are implementation details _(that is why I made all their methods private to the `neotypes` package)_.
Users only need to ensure there is an implicit instance in scope for their chosen effect and stream types.

For **Async** it was good to only require a callback because it was more flexible. And, in any case, they do not provide direct support for `CompletionStage` but rather plain callbacks.

However, a reactive stream is a very different story. Your implementation is eager, it will only work for something like `AkkaStreams`, it won't work for `fs2` for example.
I know that all streaming libraries we want to support provide a dedicated and optimized way to transform a `Publisher` into their corresponding implementations.
And those are very [complex](https://github.com/functional-streams-for-scala/fs2/blob/main/reactive-streams/src/main/scala/fs2/interop/reactivestreams/StreamSubscriber.scala) and trying to just reimplement them would be a waste.

Reply:
@BalmungSan Thanks for the reply!